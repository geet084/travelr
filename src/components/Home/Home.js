import React, { Component } from 'react';
import CountUp from 'react-countup';
import { DateForm } from '../../containers';
import Collapsible from 'react-collapsible';

class Home extends Component {
  constructor() {
    super();
    this.state = {
      time: 0,
      today: '',
    }
  }

  componentDidMount = () => {
    this.calculateDateAndTime();
  }

  calculateDateAndTime = () => {
    const now = Date.now() / 1000;
    let time = now - (this.props.time / 1000);
    const date = new Date();
    date.setDate(date.getDate() - 1);
    if (time === now) time = 0;

    this.setState({ time, today: date.toDateString() });
  }

  counter = (start, stop, decimal = 0) => {
    return (
      <CountUp
        start={start} end={stop} decimals={decimal} duration='1000000' delay={0} useEasing={false} separator=','>
        {({ countUpRef }) => <span ref={countUpRef} />}
      </CountUp>
    );
  }

  render() {
    const { url } = this.props;
    let { time, today } = this.state;
    let timeInSeconds = this.counter((time), 1000000, 1);
    let earthSpin = this.counter((time * .28), 280000, 1);
    let earthOrbit = this.counter((time * 18.5), 18500000);
    let solarSystemOrbit = this.counter((time * 124.45), 124450000);
    let galaxyMovement = this.counter((time * 70), 70000000);
    let totalMovement = this.counter((time * 213.23), 213230000);

    return (
      <div className='home'>
        <img className='apod-img' src={url} alt="apod" />
        <header>
          <div className="prompt">
            <h3 className="test">How far do you think you've traveled today?</h3>
            <h3 className="test">How far have you traveled this week?</h3>
            <h3 className="test">Enter a date to find out</h3>
          </div>
          <DateForm className="test" key={today} today={today} />
        </header>
        <Collapsible className="test" trigger="Show More">
        
        <div className='stats'>
          <h4 className="test">HOW FAST YOU ARE MOVING RIGHT NOW</h4>
          <h6 className="test">distance from new york to san francisco: <span>2,569 miles</span></h6>
          <h6 className="test">distance from san francisco to paris: <span>5,560 miles</span> </h6>
          <h6 className="test">distance from new york to australia: <span>10,512 miles</span></h6>
          <h6 className="test">circumference of the earth: <span>24,901 miles</span></h6>
          <h6 className="test">distance to the moon: <span>238,900 miles</span></h6>
        </div>
        <section className="details">
          <div className="title">
            <p className="test">elapsed time since you arrived on this page in seconds</p>
            <p className="test">Earth spinning miles per second</p>
            <p className="test">Earth orbiting the Sun miles per second</p>
            <p className="test">Solar system orbiting in the galaxy miles per second</p>
            <p className="test">Milky Way moving towards Andromeda miles per second</p>
            <p className="test">Your total speed right now in miles per second</p>
          </div>
          <div className="counter-info">
            <p className="test">{timeInSeconds}</p>
            <p className="test">{earthSpin}</p>
            <p className="test">{earthOrbit}</p>
            <p className="test">{solarSystemOrbit}</p>
            <p className="test">{galaxyMovement}</p>
            <p className="test">{totalMovement}</p>
          </div>
          </section>
        </Collapsible>
      </div>
    )
  }
}

export default Home;