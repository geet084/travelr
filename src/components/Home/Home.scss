.home {
  align-items: center;
  display: flex;
  justify-content: center;
  flex-direction: column;
  color: white;
  margin-left: 200px;
  margin: 0 auto;
}

.test {
  background-color: rgba(0,0,0,0.7);
  box-shadow: 0px 0px 5px 5px rgba(0,0,0,0.7);
}

.apod-img {
  height: 100%;
  position: fixed; 
  left: 50%; 
  top: 0;
  transform: translateX(-50%);
  z-index: -1;
}

header {
  display: flex;
  justify-content: space-between;
  padding: 10px 0 0 0;
}

.details {
 display: flex;
 justify-content: center;
}

.stats {
  width: 300px;
  margin: 0 auto;
}

.stats h4, .stats h6 {
  padding: 2px;
  margin: 0;
}

.title {
  text-align: right;
  margin: 10px 5px 10px 10px;
  height: 250px;
}

.home .Collapsible {
  position: relative;
}

.extra {
  width: 390px;
}

.counter-info {
  text-align: left;
  margin: 10px 10px 10px 5px;
  height: 250px;
  min-width: 100px;
}

.prompt {
  text-align: left;
}

.input-date {
  height: 33px;
  padding-top: 5px;
}

.user-dist {
  text-align: center;
}