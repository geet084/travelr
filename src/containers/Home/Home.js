import React, { Component } from 'react';
import DateForm from '../../components/DateForm/DateForm';
import PropTypes from 'prop-types';
import moment from 'moment';
import { connect } from 'react-redux';
import UserDate from '../../components/UserDate/UserDate';
import Distances from '../../components/Distances/Distances';
import Counter from '../../components/Counter/Counter';

export class Home extends Component {
  constructor() {
    super();
    this.state = {
      today: '',
      userDate: '',
      elapsedTime: 0,
      timeInSeconds: 0,
      earthSpin: 0,
      earthOrbit: 0,
      solarSystemOrbit: 0,
      galaxyMovement: 0,
      totalMovement: 0,
    }
  }

  componentDidMount = () => {
    this.calculateDateAndTime();
  }

  calculateDateAndTime = () => {
    const { arrivalTime } = this.props;
    const now = moment();

    const elapsedTimeSinceArrival = now.diff(arrivalTime);
    const timeInSeconds = elapsedTimeSinceArrival / 1000

    this.setState({
      timeInSeconds,
      today: now,
      earthSpin: (timeInSeconds * .28),
      earthOrbit: (timeInSeconds * 18.5),
      solarSystemOrbit: (timeInSeconds * 124.45),
      galaxyMovement: (timeInSeconds * 70),
      totalMovement: (timeInSeconds * 213.23),
    });
  }

  updateUserDate = (inputDate) => {
    // TODO: set time to a local noon or midnight?
    const { arrivalTime } = this.props;
    const userDate = moment(inputDate, 'MM-DD-YYYY');
    const elapsedTime = moment(arrivalTime).diff(userDate);
    
    this.setState({
      userDate,
      elapsedTime
    });
  }

  render() {
    const { url } = this.props;
    let { userDate, today, timeInSeconds, earthSpin, earthOrbit, solarSystemOrbit, galaxyMovement, totalMovement, } = this.state;

    const hasDate = today !== ''

    const days = hasDate ? today.diff(userDate, 'days') : 0
    const hours = hasDate ? today.diff(userDate, 'hours') - (days * 24) : 0
    const todaysDate = hasDate ? today.format('llll') : '';

    const total = ((days + (hours / 24)) * 18424000);
    const num = <Counter start={(total + totalMovement)} end={(total + totalMovement + 213230000)} decimal={1} />;
    
    return (
      <div className='home'>
        <img className='apod-img' src={url} alt="apod" />
        <header>
          <div className="prompt">
            <h3>How far do you think you've traveled today?</h3>
            <h3>How far have you traveled this week?</h3>
            <h3>Today is: {todaysDate}</h3>
            <h3>
              Enter a date to find out:
              <DateForm key={userDate} updateUserDate={this.updateUserDate} />
            </h3>
            {userDate !== '' && <UserDate userDate={userDate.format('llll')} days={days} hours={hours} num={num} />}
          </div>
        </header>
        {
          this.state.userDate !== '' &&
          <Distances
            timeInSeconds={timeInSeconds}
            earthSpin={earthSpin}
            earthOrbit={earthOrbit}
            solarSystemOrbit={solarSystemOrbit}
            galaxyMovement={galaxyMovement}
            totalMovement={totalMovement}
          />
        }
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  arrivalTime: state.arrivalTime,
})

Home.propTypes = {
  arrivalTime: PropTypes.object,
  url: PropTypes.string,
}

Home.defaultProps = {
  arrivalTime: {},
  url: '',
}

export default connect(mapStateToProps)(Home);