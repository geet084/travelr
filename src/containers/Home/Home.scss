.home {
  align-items: center;
  display: flex;
  justify-content: center;
  flex-direction: column;
  color: white;
  margin-left: 200px;
  margin: 0 auto;
}

.apod-img {
  height: 100%;
  position: fixed;
  left: 50%;
  top: 0;
  transform: translateX(-50%);
  z-index: -1;
}

header {
  display: flex;
  justify-content: space-between;
  padding: 10px 0 0 0;
}

.prompt {
  text-align: left;
  h3 {
    background-color: rgba(0, 0, 0, 0.7);
    box-shadow: 0px 0px 5px 5px rgba(0, 0, 0, 0.7);
  }
}