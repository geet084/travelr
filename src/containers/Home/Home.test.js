import React from 'react';
import { shallow } from 'enzyme';
import { Home } from './Home';
import moment from 'moment';

describe('Home', () => {
  let wrapper;
  const mockArrivalTime = moment('Mon, 1 Oct 2018 9:00 AM', 'ddd, DD MMM YYYY HH:mm:ss ZZ');
  moment.now = jest.fn().mockReturnValue(mockArrivalTime);

  beforeEach(() => {
    const mockProps = {
      url: 'http://someurl',
      arrivalTime: {_d: '1538373600000'},
    };

    wrapper = shallow(<Home {...mockProps} />);
  });

  describe('initial state', () => {
    it('should have an initial default state', () => {
      const mockProps = {
        url: 'http://someurl',
        arrivalTime: mockArrivalTime,
      };
      wrapper = shallow(<Home {...mockProps} />, { disableLifecycleMethods: true });

      const expected = {
        today: '',
        userDate: '',
        elapsedTime: 0,
        elapsed: {
          timeInSeconds: 0,
          earthSpin: 0,
          earthOrbit: 0,
          solarSystemOrbit: 0,
          galaxyMovement: 0,
          totalMovement: 0
        }
      };

      expect(wrapper.state()).toEqual(expected);
    });

    it('should match initial snapshot', () => {
      const mockProps = {
        url: 'http://someurl',
        arrivalTime: { _d: '0'},
      };
      wrapper = shallow(<Home {...mockProps} />, { disableLifecycleMethods: true });

      expect(wrapper).toMatchSnapshot();
    });

    it('call calculateDateAndTime when component mounds', () => {
      const mockProps = {
        url: 'http://someurl',
        arrivalTime: mockArrivalTime,
      };
      wrapper = shallow(<Home {...mockProps} />, { disableLifecycleMethods: true });
      wrapper.instance().calculateDateAndTime = jest.fn();

      wrapper.instance().componentDidMount();

      expect(wrapper.instance().calculateDateAndTime).toHaveBeenCalledTimes(1);
    });
  });

  describe('calculateDateAndTime', () => {
    it('should calculate date and time ', () => {
      const expected = {
        today: mockArrivalTime._d,
        elapsed: {
          timeInSeconds: 32400,
          earthSpin: 9072,
          earthOrbit: 599400,
          solarSystemOrbit: 4032180,
          galaxyMovement: 2268000,
          totalMovement: 6908652
        }
      };

      wrapper.instance().calculateDateAndTime()

      expect(wrapper.state('timeInSeconds')).toEqual(expected.timeInSeconds);
      expect(wrapper.state('today')._d).toEqual(expected.today);
      expect(wrapper.state('earthSpin')).toEqual(expected.earthSpin);
      expect(wrapper.state('earthOrbit')).toEqual(expected.earthOrbit);
      expect(wrapper.state('solarSystemOrbit')).toEqual(expected.solarSystemOrbit);
      expect(wrapper.state('galaxyMovement')).toEqual(expected.galaxyMovement);
      expect(wrapper.state('totalMovement')).toEqual(expected.totalMovement);
    });
  });

  describe('updateUserDate', () => {
    it('should update state with the user entered date and calculated elapsed time', () => {
      const mockDate = moment('09-21-2018', 'MM-DD-YYYY');
      const expected = {
        userDate: mockDate._d,
        elapsedTime: 864000000,
      };

      wrapper.instance().updateUserDate(mockDate);

      expect(wrapper.state('userDate')._d).toEqual(expected.userDate);
      expect(wrapper.state('elapsedTime')).toEqual(expected.elapsedTime);
    });
  });
});