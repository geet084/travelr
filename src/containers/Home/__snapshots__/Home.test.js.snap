// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home initial state should match initial snapshot 1`] = `
<div
  className="home"
>
  <img
    alt="apod"
    className="apod-img"
    src="http://someurl"
  />
  <header>
    <div
      className="prompt"
    >
      <h3
        className="test"
      >
        How far do you think you've traveled today?
      </h3>
      <h3
        className="test"
      >
        How far have you traveled this week?
      </h3>
      <h3
        className="test"
      >
        Today is: 
      </h3>
      <h3
        className="test input-date"
      >
        Enter a date to find out:
        <DateForm
          className="test"
          key=""
          updateUserDate={[Function]}
        />
      </h3>
    </div>
  </header>
</div>
`;
